js/logic.js

```diff
@@ -0,0 +1,862 @@
+   *
+   *   a function:
+   *     The event matches if detailPredicate(event.details) returns true.
+   *
+   * @param {string} ns
+   * @param {string} type
+   * @param {object|function} detailPredicate
+   */
+  logic.match = function(ns, type, detailPredicate) {
+    return new LogicMatcher(
+      LogicMatcher.normalizeMatchArgs(ns, type, detailPredicate));
+  }
+
+
+  function MismatchError(matcher, event) {
+    this.matcher = matcher;
+    this.event = event;
+  }
+
+  MismatchError.prototype = Object.create(Error.prototype, {
+    constructor: { value: MismatchError },
+    toString: { value: function() {
+      if (this.matcher.not) {
+        return 'MismatchError: expected ' + this.event +
+          ' to not occur (failIfMatched ' + this.matcher + ').';
+      } else {
+        return 'MismatchError: expected ' + this.event +
+          ' to match ' + this.matcher + '.';
+      }
+    }}
+  });
+
+
+  /**
+   * This is the object returned from `logic.match`. It acts as a Promise that
+   * resolves when a matching event has been logged.
+   */
+  function LogicMatcher(opts) {
+    this.matchedLogs = opts.prevMatcher ? opts.prevMatcher.matchedLogs : [];
+    this.capturedLogs = [];
+    this.ns = opts.ns;
+    this.type = opts.type;
+    this.detailPredicate = opts.detailPredicate;
+    this.failOnMismatchedDetails = true;
+    this.not = opts.not;
+    this.timeoutMS = 2000;
+    this.resolved = false;
+
+    logic.defineScope(this, 'LogicMatcher');
+
+    var prevPromise = opts.prevPromise || Promise.resolve();
+
+    if (this.not) {
+      this.promise = prevPromise.then(() => {
+        this.capturedLogs.some((event) => {
+          if ((!this.ns || event.namespace === this.ns) &&
+              event.matches(this.type, this.detailPredicate)) {
+            throw new MismatchError(this, event);
+          }
+        });
+      });
+    } else if (this.type) {
+      this.promise = new Promise((resolve, reject) => {
+        // Once any previous match has been resolved,
+        // subscribe to a following match.
+        prevPromise.then(() => {
+
+          var timeoutId = setTimeout(() => {
+            reject(new Error('LogicMatcherTimeout: ' + this));
+          }, this.timeoutMS);
+
+          // Promise chains have "dead spots" in between resolution
+          // callbacks. For instance:
+          //                 [promise1.then]      [promise2.then]
+          //    other events could be logged --^
+          //
+          // We could miss those events in the middle by just setting
+          // up a new listener for each LogicMatcher. Instead, since
+          // every matcher has a pointer to its prevMatcher, we can
+          // just grab the missing logs from there.
+          var removeMatchListener = (event) => {
+            this.resolved = true;
+            this.capturedLogs = []; // Extra events will go here.
+            setTimeout(function() {
```

Comment:
I would propose cleaning this up somewhat.  Specifically, I understand the goal to be to avoid the dead spot by using setTimeout to ensure that we only remove the listener after any chained LogicMatcher has had a chance to look at our capturedLogs before it starts listening for logs itself.  setTimeout makes sense for this because it, at least currently, will only fire after the micro-task queue has been entirely drained/processed.  Which is important because if we just did Promise.resolve().then() but there end up being some deferrals in the queue somehow, we could end up removing things before the promise actually gets a chance to run.

However, as-implemented, this still seems potentially race-prone.  I think the root of the design choice here is that since we aren't requiring the caller to cap/terminate the LogicMatcher chain that setTimeout is our best option.  But the "prevMatcher" mechanism is already an explicit internal contract.  We can have the calls to LogicMatcher.match/failIfMatched explicitly tell the prevMatcher instance that they need it to keep its logs around until they are ready.  And then, since they're reaching in to the prevMatcher, they can also explicitly trigger the event listener removal.

This has a few advantages:
* We listen to the logs for exactly as long as we need to listen, potentially avoiding some pathological memory explosions.
* LogicMatcher.match/failIfMatched can be made to throw an exception if the LogicMatcher has already transitioned to this completion state where it has added and removed its matchFn.  While it would be idiomatically weird for test code to try and chain another matcher like this, and direct use of a fresh LogicMatcher from logic.match() would encounter similar problems, the error might help save people a lot of hassle.